applications/frontier/src/frontier_handle_rate.erl

```diff
@@ -0,0 +1,330 @@
+             'true' -> wh_json:set_value(<<"Name">>, Entity, S2);
+             'false' -> S2
+         end,
+    wh_json:set_value(Type, S3, Acc).
+
+-spec make_deny_rates(ne_binary(), boolean(), ne_binaries()) -> wh_json:objects().
+make_deny_rates(Entity, IncludeRealm, MethodList) ->
+    Rates = deny_rates_for_entity(Entity, MethodList),
+    Realm = frontier_utils:extract_realm(Entity),
+    case Entity =/= Realm
+         andalso IncludeRealm
+    of
+        'true' -> Rates ++ deny_rates_for_entity(Realm, MethodList);
+        'false' -> Rates
+    end.
+
+%% Kamailio expects -1 for unkown entities
+-spec deny_rates_for_entity(ne_binary(), ne_binaries()) -> list().
+deny_rates_for_entity(Entity, MethodList) ->
+    lists:flatmap(fun (M) -> construct_records(M, Entity, -1, -1) end, MethodList).
+
+-spec construct_records(ne_binary(), ne_binary(), ne_binary(), ne_binary()) -> wh_json:objects().
+construct_records(Method, Entity, RPM, RPS) ->
+    {Name, Type} = case binary:split(Entity, <<"@">>) of
+                       [User, _] -> {User, <<"device">>};
+                       [Realm] -> {Realm, <<"realm">>}
+                   end,
+    RPMObject = wh_json:from_list([{<<"type">>, Type}
+                                   ,{?MINUTE, RPM}
+                                  ]),
+    RPSObject = wh_json:from_list([{<<"type">>, Type}
+                                   ,{?SECOND, RPS}
+                                  ]),
+    Record = wh_json:from_list([{<<"id">>, 'undefined'}
+                                ,{<<"key">>, [Name, Method]}
+                               ]),
+    lists:map(fun (Obj) -> wh_json:set_value(<<"value">>, Obj, Record) end, [RPMObject, RPSObject]).
+
+
+-spec fetch_rates_from_sys_config(ne_binaries(), ne_binary(), ne_binaries()) -> 'ok'.
+fetch_rates_from_sys_config(_, _, []) ->
+    lager:info("sysconfig: Empty request - empty response"),
+    [];
+fetch_rates_from_sys_config(Entity, Type, MethodList) ->
+    Section = case Type of
+                  <<"device">> -> <<"device">>;
+                  <<"realm">> -> <<"account">>
+              end,
+    AllRates = whapps_config:get(?APP_NAME, <<"rate_limits">>),
+    TargetRates = wh_json:get_value(Section, AllRates, wh_json:new()),
+    lists:foldl( fun (Method, Acc) ->
+                      RPM = wh_json:get_value([?MINUTE, Method], TargetRates),
+                      RPS = wh_json:get_value([?SECOND, Method], TargetRates),
+                      case RPM =:= 'undefined' orelse RPS =:= 'undefined' of
+                          'true' -> Acc;
+                          'false' -> construct_records(Method, Entity, RPM, RPS) ++ Acc
+                      end
+                 end, [], MethodList).
+
+-spec fetch_rates(ne_binary(), boolean(), ne_binaries(), ne_binary()) -> wh_json:objects().
+fetch_rates(_, _, [], _) ->
+    lager:info("document: Empty request - empty response"),
+    [];
+fetch_rates([], _, _, _) ->
```

Comment:
Dialyzer reports this clause as being uncallable and I agree. I only see this arg being filled as a result of a call to `build_list_of_querynames/2` which always returns a non-empty list.